lib/exercises.rb

```diff
@@ -1,19 +1,62 @@
 
 # This method will return an array of arrays.
 # Each subarray will have strings which are anagrams of each other
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n * m(logm)), n = num of strings, m = length of string
+# Space Complexity: O(n) where n is the number of anagram groups
 
 def grouped_anagrams(strings)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+  # assumptions:
+  # all strings 3 letters
+
+  strings_hash = Hash.new
+  strings.each do |string|
+    key = string.split('').sort #O(n) + O(log n)
+    # split into letters, alphabetize, use as key
+    if strings_hash[key]
+      strings_hash[key].push(string)
+    else
+      strings_hash[key] = [string]
+    end
+  end
+
+  return strings_hash.values
+
 end
 
 # This method will return the k most common elements
 # in the case of a tie it will select the first occuring element.
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n), where n is number of elements
+# Space Complexity: O(n)
 def top_k_frequent_elements(list, k)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+
+  # assumptions:
+  # list is integers, non-empty - do account for 1 num arr
+  nums_hash = Hash.new
+  list.each do |num|
+    if nums_hash[num]
+      nums_hash[num] += 1
+    else
+      nums_hash[num] = 1
+    end
+  end
+
+  most_frequent = nums_hash.values.max(k)
```

Comment:
You will want to check and verify the time complexity of `max(k)` and see what's the time complexity here.